Stowaways now get a 2-use maintenance ID

KoboldCommando opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

# About The Pull Request

Stowaways will now spawn with a fake ID.  It's good for 2 uses on maintenance doors.  It has to be used directly and takes time to use.  After using it twice it's worthless, but it can still be used for display so you don't show up as Unknown.

It also fixes a bug where stowaways could spawn inside locked lockers, sometimes unable to escape.

It additionally removes a double parent call in /obj/item/card/id/examine

## Why It's Good For The Game

Stowaway is fun for a variety of reasons, but the likelihood of being locked in a maintenance room without the tools needed to escape is very high.  Being able to open a couple maint doors will alleviate a huge chunk of these cases.  You can still wind up stuck, or in areas where maint access won't help, but that's part of the quirk.  Being stuck on the station in camera range isn't as bad as a dark maintenance room, and this solution preserves a bit of the unpredictability and spice of the random spawns.

This also gives stowaways a little more identification, so bloodthirsty security players don't immediately bash the Unknown.

It also forms a base if we decide to implement a temporary ID system in the future, similar cards could be printable by heads and given out for ~10 uses of departmental doors.

## Testing Photographs and Procedure

![image](https://user-images.githubusercontent.com/31165061/198411377-820f8e72-ff95-473a-914f-efe07605ecf5.png)
![image](https://user-images.githubusercontent.com/31165061/198411387-80923576-0cc6-4c3a-a552-05ebfb49b317.png)
![image](https://user-images.githubusercontent.com/31165061/198411401-3c534cbc-6b51-4f54-bba4-5dcc66573d05.png)
![image](https://user-images.githubusercontent.com/31165061/198411422-65acaedd-f7a7-4a4f-aa48-f1032db6a156.png)
![image](https://user-images.githubusercontent.com/31165061/198411463-ac2e147f-c7f7-4d84-8db8-1ea221851673.png)


## Changelog

:cl:
add: Stowaways now spawn with a limited-use fake ID card, good for any maintenance door!
fix: Stowaways should no longer occasionally spawn in locked lockers, unable to escape.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

nednaZ commented:
I'm not sure about this one.
A two use ID isn't going to be enough to get out of many maintenance locations and it does nullify a lot of the penalty of taking this quirk in the first place.

KoboldCommando commented:
> I'm not sure about this one. A two use ID isn't going to be enough to get out of many maintenance locations and it does nullify a lot of the penalty of taking this quirk in the first place.

My interpretation is that the loss of an ID for day to day operations, not being on the manifest, and starting in a random location were the primary features of the perk.

Showing up in maintenance without the means to escape is FAR more prevalent here than it was on goon, due to our larger maintenance tunnels, and is far less fun.  I ran the perk for a while, and it was enjoyable to puzzle out how to escape, but more often than not one room wasn't enough.  Two or three rooms, however, would almost certainly have the tools to at least bust through a wall into the main hallway and then you're good.

You still have the risk of appearing in non-maintenance sealed rooms as well.  But the difference there is that most of them have cameras, making it MUCH more likely for you to be found and rescued.

KoboldCommando commented:
I've added in a fix for the locked locker issue.  Stowaways will no longer pop out of locked crates or lockers, so they can't show up in the toxins or firing range crates or similar spots.

I tested this by duplicating the secure toxins crate a zillion times and then repeatedly using atom proccall on myself to call forceMove with a proc argument of /proc/get_unlocked_closed_locker (the underlying proc used by stowaway to get a locker)

KoboldCommando commented:
Also added a hint to the description of the cards that you have to use them directly rather than just bumping doors.

nednaZ commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -857,3 +857,34 @@ update_label("John Doe", "Clowny")
+	uses -= 1
+	if(uses == 1)
+		icon_state = "counterfeit_torn"
+		desc = "Definitely a legitimate ID card and not a piece of notebook paper with a magnetic strip drawn on it. You'd have to stuff this in a card reader by hand for it to work. It's falling apart!"
+	if(uses < 1)
+		icon_state = "counterfeit_torn2"
+	desc = "Definitely a legitimate ID card and not a piece of notebook paper with a magnetic strip drawn on it. You'd have to stuff this in a card reader by hand for it to work. It's too shredded to fit in a scanner!"
```
This can be on the examine proc rather than here.
Look into how other examine overrides handle it, you can have part of the examination appended on rather than changing the entire thing.

nednaZ commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -857,3 +857,34 @@ update_label("John Doe", "Clowny")
+	if(uses == 1)
+		icon_state = "counterfeit_torn"
+		desc = "Definitely a legitimate ID card and not a piece of notebook paper with a magnetic strip drawn on it. You'd have to stuff this in a card reader by hand for it to work. It's falling apart!"
+	if(uses < 1)
```
if(!uses)

nednaZ commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -857,3 +857,34 @@ update_label("John Doe", "Clowny")
+
+/obj/item/card/id/fake_card/proc/used()
+	uses -= 1
+	if(uses == 1)
```
I'd also say a switch statement is best.
If(2), (1), (0)

nednaZ commented on monkestation/code/__HELPERS/unsorted.dm:
```diff
@@ -3,7 +3,7 @@
 	var/turf/object_location
 	for(var/obj/structure/closet/find_closet in world)
-		if(!istype(find_closet,/obj/structure/closet/secure_closet))
+		if(!find_closet.locked)
```
Note that this will increase the number of possible locations. Unlocked head lockers will now be an option.

KoboldCommando replied:
Yeah it's kind of a shift.  Fewer locked boxes, but more unlocked boxes.  I think it's probably fine because it's just more incentive to lock your lockers.

nednaZ commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -857,3 +857,34 @@ update_label("John Doe", "Clowny")
+	slot_flags = ITEM_SLOT_ID
+	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 100, "acid" = 100, "stamina" = 0)
+	resistance_flags = FIRE_PROOF | ACID_PROOF
+	registered_name = "Nohbdy"
```
Misspelling
```suggestion
	registered_name = "Nobody"
```

KoboldCommando replied:
Actually intentional, at least in the version of The Odyssey I read they used that spelling.  "I have Nohbdy's ID card!"  Hopefully that never shows up in game anyway though.

nednaZ commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -857,3 +857,34 @@ update_label("John Doe", "Clowny")
+	registered_name = new_name
+	name = "[new_name]'s \"ID Card\""
+
+/obj/item/card/id/fake_card/proc/used()
```
I'd suggest having a single icon_state update after the use.
You'll need to update the icon name to be counterfeit_torn1 and counterfeit_torn0 however.

"icon_state = counterfeit_torn[uses]"

KoboldCommando replied:
I feel like doing it dynamically could potentially result in problems?  I've been thinking of doing a followup to allow creation of temporary IDs with more than 2 initial uses.  My intent was to have it appear normal until your last use, and then have a "broken" state.

nednaZ replied:
If you have further plans for these cards beyond this system, then it's fine.
You'll just need a switch statement with X to INFINITY to handle higher numbers to make sure that there aren't errors.

nednaZ reviewed (changes requested):
Does it check for a welded or powered down door before the fake card is used?
May want to do that so that it isn't wasting a "charge" in case the door wouldn't work anyway.

nednaZ commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -857,3 +857,44 @@ update_label("John Doe", "Clowny")
+	return //no accounts on fake cards
+
+/obj/item/card/id/fake_card/examine(mob/user)
+	..()
+	. = ..()
```
Double parent call.

KoboldCommando commented:
Fixed the examine text, and also the parent examine call which had a similar double parent call.

Also fixed the switch statement to handle uses higher than 2 rather than just letting them fall off the end.

KoboldCommando commented:
Examine text ACTUALLY fixed now, fake IDs have no message about alt-clicking.  I tested all other ID subtypes and they still have them.

nednaZ commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -1198,6 +1200,28 @@
+	var/obj/item/card/id/fake_card/fake = card
+	if(fake.uses)
+		if(check_access_list(fake.access))
+			visible_message("<span class='warning'>[user] starts fumbling at [src] with a piece of paper!</span>")
```
Should have a message sent to the user as well, stating that "you start swiping the card at the door" or somethijng along those lines.
I believe self_message is an argument for visible_message

nednaZ commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -857,3 +858,48 @@ update_label("John Doe", "Clowny")
 		to_chat(user, "You upgrade your [idcard] with the [name].")
 		log_id("[key_name(user)] added access to '[idcard]' using [src] at [AREACOORD(user)].")
 		qdel(src)
+/obj/item/card/id/fake_card //not a proper ID but still shares a lot of functions
```
Add a new line between the previous definition and the new one.

nednaZ reviewed (changes requested):
Looks just about ready to go.

KoboldCommando commented:
Changes made!

nednaZ reviewed (approved):
Looks good to me.
